03-intro-to-python.Rmd

```diff
@@ -1,9 +1,298 @@
+
+In Python, the equivalent if statement looks like:
+
+```{python}
+x = 1
+
+if x > 0:
+    print("x is positive")
+else:
+    print("x is negative")
+```
+
+
+### Data Structures
+
+There are 4 different data storage formats, or data structures, in Python: lists, tuples, dictionaries, and sets
+
+#### Lists
+
+Python lists are created using brackets `[]`. You can add elements to the list through the `append()` method.
+
+```{python}
+x = [1, 2, 3]
+x.append(4) # add 4 to the end of list
+
+print("x is", x)
+#> x is [1, 2, 3, 4]
+```
+
+
+You can index into lists with integers using brackets `[]`, but note that indexing is 0-based.
+
+```{python}
+x = [1, 2, 3]
+
+x[0]
+#> 1
+x[1]
+#> 2
+x[2]
+#> 3
+```
+
+
+Negative numbers count from the end of the list.
+
+```{python}
+x = [1, 2, 3]
+
+x[-1]
+#> 3
+x[-2]
+#> 2
+x[-3]
+#> 1
+```
+
+You can slice ranges of lists using the : inside brackets. Note that the slice syntax is not inclusive of the end of the slice range.
+
+```{python}
+x = [1, 2, 3, 4, 5, 6]
+x[0:2] # get items at index positions 0, 1
+#> [1, 2]
+x[1:]  # get items from index position 1 to the end
+#> [2, 3, 4, 5, 6]
+x[:-2] # get items from beginning up to the 2nd to last.
+#> [1, 2, 3, 4]
+x[:]   # get all the items
+#> [1, 2, 3, 4, 5, 6]
+```
+
+
+#### Tuples
+
+Tuples behave like lists, but are constructued using `()`, instead of `[]`. 
+
+```{python}
+x = (1, 2) # tuple of length 2
+type(x)
+#> <class 'tuple'>
+len(x)
+#> 2
+x
+#> (1, 2)
+
+x = (1,) # tuple of length 1
+type(x)
+#> <class 'tuple'>
+len(x)
+#> 1
+x
+#> (1,)
+
+x = 1, 2 # also a tuple
+type(x)
+#> <class 'tuple'>
+len(x)
+#> 2
+
+x = 1, # beware a single trailing comma! This is a tuple!
+type(x)
+#> <class 'tuple'>
+len(x)
+#> 1
+```
+
+#### Dictionaries
+
+Dictionaries are data structures where you can retrieve items by name. They can be created using syntax like {key: value}.
+
+```{python}
+d = {"key1": 1,
+     "key2": 2}
+
+d["key1"] 
+#> 1
+d["key3"] = 3
+d 
+#> {'key1': 1, 'key2': 2, 'key3': 3}
+```
+
+#### Sets 
+
+Sets are used to track unique items, and can be constructed using `{val1, val2}`.
+
+```{python}
+s = {1, 2, 3}
+
+type(s)
+#> <class 'set'>
+s
+#> {1, 2, 3}
+```
+
+### Iteration with for loops
+ 
+The `for` statement in Python is similar to the `for` loop in R. It can be used to iterate over any kind of data structure.
+
+```{python}
+for x in [1, 2, 3]:
+  print(x)
+#> 1
+#> 2
+#> 3
+```
+
+### Functions
+
+Python functions are defined with the def statement. The syntax for specifying function arguments and default values is very similar to R.
+
+```{python}
+def my_function(name = "World"):
+  print("Hello", name)
+
+my_function()
+#> Hello World
+my_function("Friend")
+#> Hello Friend
+```
+
+The equivalent R code would be
+
+```{r}
+my_function <- function(name = "World") {
+  cat("Hello", name, "\n")
+}
+
+my_function()
+#> Hello World
+my_function("Friend")
+#> Hello Friend
+```
+
+
+### Classes and Object Oriented Programming (OOP)
```

Comment:
Should we add something here about why OOP is useful or interesting, e.g., "co-locating" data and methods? Inheritance we should maybe save for a later chapter?

Reply:
I don't know if talking about OOP topics like inheritance or encapsulation is too advanced for the students. I'll leave this up to @cansavvy

Reply:
I think it would be good to introduce the students to this concept, but I'm not sure if we want to do it so soon. This should probably be its own chapter. 

OOP is something a bit foreign for R users and something R can't do very well. So I would add that to the start comparison part at the beginning of this chapter in very general terms (i'll put a suggestion for how to do that) but then we can dive into the details later